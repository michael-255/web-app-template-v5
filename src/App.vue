<script setup lang="ts">
import useLogger from '@/composables/useLogger'
import DB from '@/services/Database'
import { appDescription } from '@/shared/constants'
import { errorIcon } from '@/shared/icons'
import useSettingsStore from '@/stores/settings'
import { colors, useMeta, useQuasar } from 'quasar'
import { onMounted, onUnmounted } from 'vue'
import { RouterView } from 'vue-router'

/**
 * Sets up meta tags and links for the app. These are primarily for the favicons and manifest.
 * Do NOT overwrite these specific properties in another useMeta().
 */
useMeta({
    meta: {
        description: { name: 'description', content: appDescription },
        charset: { charset: 'UTF-8' },
        viewport: {
            name: 'viewport',
            content: 'width=device-width, initial-scale=1.0',
        },
        themeColor: { name: 'theme-color', content: `${colors.getPaletteColor('primary')}` },
    },

    // link: {
    //     manifest: {
    //         rel: 'manifest',
    //         href: `${import.meta.env.BASE_URL}manifest.json`,
    //     },
    //     appleTouchIcon: {
    //         rel: 'apple-touch-icon',
    //         sizes: '180x180',
    //         href: `${import.meta.env.BASE_URL}apple-touch-icon.png`,
    //     },
    //     favicon32: {
    //         rel: 'icon',
    //         type: 'image/png',
    //         sizes: '32x32',
    //         href: `${import.meta.env.BASE_URL}favicon-32x32.png`,
    //     },
    //     favicon16: {
    //         rel: 'icon',
    //         type: 'image/png',
    //         sizes: '16x16',
    //         href: `${import.meta.env.BASE_URL}favicon-16x16.png`,
    //     },
    // },

    noscript: {
        default:
            'Your browser does not support JavaScript or has it disabled. Please enable JavaScript in your web browser settings or white-list our domain in your JavaScript blocker for the best experience.',
    },
})

const notify = useQuasar().notify
const { log } = useLogger()
const settingsStore = useSettingsStore()

const settingsSubscription = DB.liveSettings().subscribe({
    next: (liveSettings) => {
        settingsStore.settings = liveSettings
    },
    error: (error) => {
        log.error('Error loading live settings', error)
    },
})

onMounted(async () => {
    try {
        const settings = await DB.initSettings()
        log.silentDebug('Settings initialized', settings)
    } catch (error) {
        // This isn't saving the error since it could be a DB or logger failure
        notify({ message: 'Error initializing settings', icon: errorIcon, color: 'negative' })
    }

    try {
        const logsPurged = await DB.purgeLogs()
        log.silentDebug('Expired logs purged', { logsPurged })
    } catch (error) {
        log.error('Error purging expired logs', error)
    }
})

onUnmounted(() => {
    settingsSubscription.unsubscribe()
})
</script>

<template>
    <RouterView />
</template>
