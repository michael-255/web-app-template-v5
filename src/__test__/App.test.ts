import { createTestingPinia } from '@pinia/testing'
import { flushPromises, shallowMount } from '@vue/test-utils'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import App from '../App.vue'

// Required for these spys to work with vi.mock()
const spys = vi.hoisted(() => ({
    useMeta: vi.fn(),
    notify: vi.fn(),
    initSettings: vi.fn(),
    purgeLogs: vi.fn(),
    subscribeSettings: vi.fn(),
    logSilentDebug: vi.fn(),
    logError: vi.fn(),
}))

vi.mock('quasar', () => {
    // Workaround to get the dialog plugin to stop breaking tests
    const dialogPlugin = () => {
        return {
            dialogRef: {},
            onDialogHide: () => {},
            onDialogOK: () => {},
            onDialogCancel: () => {},
        }
    }
    dialogPlugin.emits = ['ok', 'hide']

    return {
        useDialogPluginComponent: dialogPlugin,
        useMeta: spys.useMeta,
        useQuasar: () => ({
            notify: spys.notify,
        }),
        colors: {
            getPaletteColor: vi.fn(), // So useMeta test won't break
        },
    }
})

vi.mock('../services/Database.ts', () => ({
    default: {
        initSettings: spys.initSettings,
        purgeLogs: spys.purgeLogs,
        liveSettings: () => ({
            subscribe: spys.subscribeSettings,
        }),
    },
}))

vi.mock('../composables/useLogger.ts', () => ({
    default: () => ({
        log: {
            silentDebug: spys.logSilentDebug,
            error: spys.logError,
        },
    }),
}))

/**
 * Helper that returns the wrapped component. For some tests the wrapper just needs to be
 * constructed, but isn't needed beyond that.
 */
function buildAppWrapper() {
    return shallowMount(App, {
        global: {
            plugins: [
                createTestingPinia({
                    createSpy: vi.fn(),
                    initialState: {
                        settings: [],
                    },
                }),
            ],
        },
    })
}

describe('App', () => {
    beforeEach(() => {
        vi.resetAllMocks()
    })

    it('should match the snapshot', () => {
        const wrapper = buildAppWrapper()
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('should call useMeta()', () => {
        buildAppWrapper()
        expect(spys.useMeta).toHaveBeenCalledWith(
            expect.objectContaining({
                meta: expect.any(Object),
            }),
        )
    })

    it('should subscribe to live settings', async () => {
        buildAppWrapper()
        await flushPromises()
        expect(spys.subscribeSettings).toHaveBeenCalled()
    })

    it('should initialize settings on mount', async () => {
        const settings = [{ key: 'setting', value: true }]
        spys.initSettings.mockResolvedValueOnce(settings)
        buildAppWrapper()
        await flushPromises()
        expect(spys.initSettings).toHaveBeenCalled()
        expect(spys.logSilentDebug).toHaveBeenCalledWith('Settings initialized', settings)
    })

    it('should purge logs on mount', async () => {
        spys.purgeLogs.mockResolvedValueOnce(0)
        buildAppWrapper()
        await flushPromises()
        expect(spys.purgeLogs).toHaveBeenCalled()
        expect(spys.logSilentDebug).toHaveBeenCalledWith('Expired logs purged', { logsPurged: 0 })
    })

    it('should notify the user if initializing settings fails', async () => {
        spys.initSettings.mockImplementationOnce(() => {
            throw new Error('Test error')
        })
        buildAppWrapper()
        await flushPromises()
        expect(spys.notify).toHaveBeenCalledWith(
            expect.objectContaining({
                message: 'Error initializing settings',
            }),
        )
    })

    it('should error log if purge logs fails', async () => {
        const error = new Error('Test error')
        spys.purgeLogs.mockImplementationOnce(() => {
            throw error
        })
        buildAppWrapper()
        await flushPromises()
        expect(spys.logError).toHaveBeenCalledWith('Error purging expired logs', error as Error)
    })
})
