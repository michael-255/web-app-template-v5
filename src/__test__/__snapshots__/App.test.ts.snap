// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should match the snapshot 1`] = `"<router-view-stub name=\\"default\\"></router-view-stub>"`;
